extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-sqlserver/schema/001-create-schema.sql

```diff
@@ -0,0 +1,558 @@
+    CHECK CONSTRAINT [CK_proxy_encryption_method];
+ALTER TABLE [guacamole_connection]
+    ADD CONSTRAINT [DF_guacamole_connection_failover_only] DEFAULT ((0)) FOR [failover_only];
+GO;
+
+/**
+ * The user table stores user accounts, passwords, and properties.
+ */
+SET ANSI_NULLS ON;
+SET QUOTED_IDENTIFIER ON;
+CREATE TABLE [guacamole_user](
+    [user_id] [int] IDENTITY(1,1) NOT NULL,
+    [username] [nvarchar](128) NOT NULL,
+    [password_hash] [binary](32) NOT NULL,
+    [password_salt] [binary](32) NULL,
+    [password_date] [datetime] NOT NULL,
+    [disabled] [bit] NOT NULL,
+    [expired] [bit] NOT NULL,
+    [access_window_start] [time](7) NULL,
+    [access_window_end] [time](7) NULL,
+    [valid_from] [date] NULL,
+    [valid_until] [date] NULL,
+    [timezone] [nvarchar](64) NULL,
+    [full_name] [nvarchar](256) NULL,
+    [email_address] [nvarchar](256) NULL,
+    [organization] [nvarchar](256) NULL,
+    [organizational_role] [nvarchar](256) NULL,
+
+    CONSTRAINT [PK_guacamole_user] PRIMARY KEY CLUSTERED 
+        ([user_id] ASC)
+        WITH (PAD_INDEX = OFF,
+            STATISTICS_NORECOMPUTE = OFF,
+            IGNORE_DUP_KEY = OFF,
+            ALLOW_ROW_LOCKS = ON,
+            ALLOW_PAGE_LOCKS = ON)
+        ON [PRIMARY]
+) ON [PRIMARY];
+
+/**
+ * Defaults for user table
+ */
+ALTER TABLE [guacamole_user]
+    ADD CONSTRAINT [DF_guacamole_user_disabled] DEFAULT ((0)) FOR [disabled];
+ALTER TABLE [guacamole_user]
+    ADD CONSTRAINT [DF_guacamole_user_expired] DEFAULT ((0)) FOR [expired];
+GO;
+
+/**
+ * The sharing_profile table stores profiles that allow
+ * connections to be shared amongst multiple users.
+ */
+SET ANSI_NULLS ON;
+SET QUOTED_IDENTIFIER ON;
+CREATE TABLE [guacamole_sharing_profile](
+    [sharing_profile_id] [int] IDENTITY(1,1) NOT NULL,
+    [sharing_profile_name] [nvarchar](128) NOT NULL,
+    [primary_connection_id] [int] NOT NULL,
+
+    CONSTRAINT [PK_guacamole_sharing_profile] PRIMARY KEY CLUSTERED 
+        ([sharing_profile_id] ASC)
+        WITH (PAD_INDEX = OFF,
+            STATISTICS_NORECOMPUTE = OFF,
+            IGNORE_DUP_KEY = OFF,
+            ALLOW_ROW_LOCKS = ON,
+            ALLOW_PAGE_LOCKS = ON)
+        ON [PRIMARY]
+) ON [PRIMARY];
+
+/**
+ * Foreign keys for sharing_profile table.
+ */
+ALTER TABLE [guacamole_sharing_profile]
+    WITH CHECK ADD CONSTRAINT [FK_guacamole_sharing_profile_connection] FOREIGN KEY([primary_connection_id])
+    REFERENCES [guacamole_connection] ([connection_id])
+        ON UPDATE CASCADE
+        ON DELETE CASCADE;
+ALTER TABLE [guacamole_sharing_profile]
+    CHECK CONSTRAINT [FK_guacamole_sharing_profile_connection];
+GO;
+
+/**
+ * The connection_parameter table stores parameters for
+ * connection objects.
+ */
+SET ANSI_NULLS ON;
+SET QUOTED_IDENTIFIER ON;
+CREATE TABLE [guacamole_connection_parameter](
+    [connection_id] [int] NOT NULL,
+    [parameter_name] [nvarchar](128) NOT NULL,
+    [parameter_value] [nvarchar](max) NOT NULL,

```

Comment:
Why `max`?

Reply:
max = 4000, which is less than what Postgres and MySQL use (4096).

Reply:
On recent versions of SQL Server, values up to 8000 are allowed, while a `varchar(max)` indicates the value should can grow up to 2 GB. According to https://docs.microsoft.com/en-us/sql/t-sql/data-types/char-and-varchar-transact-sql (emphasis added):

> varchar [ ( n | max ) ] Variable-length, non-Unicode string data. n defines the string length and can be a value from 1 through 8,000. **max indicates that the maximum storage size is 2^31-1 bytes (2 GB)**. The storage size is the actual length of the data entered + 2 bytes. The ISO synonyms for varchar are charvarying or charactervarying.

Not sure if the definition of this has changed between various versions of SQL Server, but assuming that's the case, that possibility may be another reason  to avoid `max`.

Are you testing primarily against SQL Server 2005 and its limit is 4000?

Reply:
Nope, I'm testing against SQLServer on Linux, or SQLServer 2017.  When I tried to use the 4096 value, I was told it was above the max.

<Shrug>?!

Reply:
Ah, actually, the issue is that I'm using nvarchar instead of varchar, because nvarchar supports storing unicode values.

Reply:
I'm still concerned about the use of `max` here, as the documentation states that this is equivalent to a maximum storage size of 2 GB, not the intended 4 KB.

If the intent is to have a `[nvarchar](4000)`, why not use `[nvarchar](4000)`?